includes/admin/misc-functions.php

```diff
@@ -249,3 +249,99 @@ function give_add_ons_feed( $feed_type = '', $echo = true ) {
 
 	return $cache;
 }
+
+/**
+ * Handle installation and connection for SendWP via ajax
+ *
+ * @since 2.9.15
+ */
+function give_sendwp_remote_install_handler () {
+
+	if ( ! current_user_can( 'manage_give_settings' ) ) {
+		wp_send_json_error( array(
+			'error' => __( 'You do not have permission to do this.', 'give' )
+		) );
+	}
+
+	include_once ABSPATH . 'wp-admin/includes/plugin-install.php';
+	include_once ABSPATH . 'wp-admin/includes/file.php';
+	include_once ABSPATH . 'wp-admin/includes/class-wp-upgrader.php';
+
+	$plugins = get_plugins();
+
+	if( ! array_key_exists( 'sendwp/sendwp.php', $plugins ) ) {
+
+		/*
+		* Use the WordPress Plugins API to get the plugin download link.
+		*/
+		$api = plugins_api( 'plugin_information', array(
+			'slug' => 'sendwp',
+		) );
+
+		if ( is_wp_error( $api ) ) {
+			wp_send_json_error( array(
+				'error' => $api->get_error_message(),
+				'debug' => $api
+			) );
+		}
+
+		/*
+		* Use the AJAX Upgrader skin to quietly install the plugin.
+		*/
+		$upgrader = new Plugin_Upgrader( new WP_Ajax_Upgrader_Skin() );
+		$install = $upgrader->install( $api->download_link );
+		if ( is_wp_error( $install ) ) {
+			wp_send_json_error( array(
+				'error' => $install->get_error_message(),
+				'debug' => $api
+			) );
+		}
+
+		$activated = activate_plugin( $upgrader->plugin_info() );
+
+	} else {
+
+		$activated = activate_plugin( 'sendwp/sendwp.php' );
+
+	}
+
+	/*
+	* Final check to see if SendWP is available.
+	*/
+	if( ! function_exists('sendwp_get_server_url') ) {
+		wp_send_json_error( array(
+			'error' => __( 'Something went wrong. SendWP was not installed correctly.', 'give' )
+		) );
+	}
+
+	wp_send_json_success( array(
+		'partner_id'      => 51154,
+		'register_url'    => sendwp_get_server_url() . '_/signup',
+		'client_name'     => sendwp_get_client_name(),
+        'client_url'      => sendwp_get_client_url(),
+		'client_secret'   => sendwp_get_client_secret(),
+		'client_redirect' => admin_url( '/edit.php?post_type=give_forms&page=give-settings&tab=emails&section=email-settings' ),
+	) );
+}
+add_action( 'wp_ajax_give_sendwp_remote_install', 'give_sendwp_remote_install_handler' );
+
+/**
+ * Handle deactivation of SendWP via ajax
+ *
+ * @since 2.9.15
+ */
+function give_sendwp_disconnect () {
+
+	if ( ! current_user_can( 'manage_shop_settings' ) ) {
```

Comment:
Is that a GiveWP function? seems like EDD or something.

Reply:
I've updated this to be more in line with GiveWP capabilities. Good catch!